sympy/stats/stochastic_process_types.py

```diff
@@ -580,18 +581,72 @@ def expectation(self, expr, condition=None, evaluate=True, **kwargs):
         raise NotImplementedError("Mechanism for handling (%s, %s) queries hasn't been "
                                 "implemented yet."%(expr, condition))
 
+
+class CommunicationClass:
```

Comment:
Why is this class needed?
If there is a whole class needed to handle complicated logic, I'd not be happy with introducing custom object types to sympy that are not immutable and not instances of `Basic`

Reply:
I can make it an inherit from `Basic` but it needs to be mutable because its construction is very "dynamic". I thought that since it is super light-weight that it shouldn't need to inherit from anything. I'll make it inherit from `Basic` but I'm pretty new to creating SymPy classes from scratch. So any suggestions on methods that I should override are greatly appreciated.

Reply:
I'm thinking of other ideas of bookkeeping properties like `.recurrent` or `.period` as a cache of Markov chain itself, which is accessible by functions like `MarkovChain.is_recurrent`, `MarkovChain.get_period`. 
And `communication_class` can just returning python set objects.

Reply:
I think it is better to have a class since each Markov Chain has a list of Communication Classes that partition the state space. Each communication class then shares properties like recurrence and period. So if we wanted to cache these properties separately, we would need a list of lists for the partition, a list for recurrence and a list for period. 

I think the code to create these Communication Classes will also be much longer and less readable.

Reply:
Are the communication classes the same as the connected components of the matrix when viewed as an adjacency graph?

For example:
```julia
In [1]: M = Matrix([[1, 0, 1], [0, 1, 0], [0, 0, 1]])                                                                                                         

In [2]: M                                                                                                                                                     
Out[2]: 
⎡1  0  1⎤
⎢       ⎥
⎢0  1  0⎥
⎢       ⎥
⎣0  0  1⎦

In [3]: M.connected_components()                                                                                                                              
Out[3]: [[0, 2], [1]]

In [4]: M.connected_components_decomposition()                                                                                                                
Out[4]: 
⎛                          ⎡⎡1  1⎤     ⎤⎞
⎜                          ⎢⎢    ⎥   𝟘 ⎥⎟
⎜PermutationMatrix((1 2)), ⎢⎣0  1⎦     ⎥⎟
⎜                          ⎢           ⎥⎟
⎝                          ⎣  𝟘     [1]⎦⎠
```

Reply:
Usually, I'd expect public classes should be nontrivial and easy to expect, so I'd hardly imagine a bunch of properties ragged together with the object while I'd expect it to be simple mathematically well-defined objects like sets or lists.

Do CommunicationClass intended to be manipulated outside of `communication_classes` function? If not, I'd rather figure out any ways to hide such implementation details internally.
So you can rather cache the CommunicationClass objects and it can solve the problem.

Reply:
> Are the communication classes the same as the connected components of the matrix when viewed as an adjacency graph?

They are similar but this seems to be very liberal with grouping them. Communication classes are only grouped together if both the row and column connect eventually. 

For example 
```python
# example 1
Matrix([[0.5, 0.5, 0], [0, 1, 0], [0, 0.5, 0.5]])
M.connected_components()  # [[0, 1, 2]]

# example 2
M = Matrix([[0, 1, 0], [0, 0, 1], [1, 0, 0]])
M.connected_components()  # [[0, 1, 2]]
```` 
For example 1:
But the communication classes are `[[0], [1], [2]]`. This is because state 1 is reachable from both states 0 and 2 but not the other way round. So none of the states communicate and they all have separate communication classes.

For example 2:
In this case they are the same. This is because state 0 can reach state 1 which can also reach state 2 which in turn reaches state 0. So every state is accessible from every other state and they all form part of the same class.

These functions look really cool. My IDE doesn't pick up these functions when I type `M.` so I never even knew they existed. My PyCharm seems to have this problem with most classes in SymPy.

Reply:
I think it's the difference between (weakly-)connected and strongly connected components:
https://github.com/sympy/sympy/blob/e975b8a7b339ee2e209f6d92e8ac58ee278ad5d9/sympy/utilities/iterables.py#L1168
https://github.com/sympy/sympy/blob/e975b8a7b339ee2e209f6d92e8ac58ee278ad5d9/sympy/utilities/iterables.py#L1045

If you partition the matrix into strongly connected components then I think that those are the communication classes:
```julia
In [30]: from sympy.utilities.iterables import connected_components                                                                                           

In [31]: from sympy.utilities.iterables import strongly_connected_components                                                                                  

In [32]: M = Matrix([[0.5, 0.5, 0], [0, 1, 0], [0, 0.5, 0.5]])                                                                                                

In [33]: V = range(3)                                                                                                                                         

In [34]: E = [(i, j) for i in range(3) for j in range(3) if M[i, j]]                                                                                          

In [35]: G = (V, E)                                                                                                                                           

In [36]: connected_components(G)                                                                                                                              
Out[36]: [[0, 1, 2]]

In [37]: strongly_connected_components(G)                                                                                                                     
Out[37]: [[1], [0], [2]]
```

Reply:
> I'd expect public classes should be nontrivial and easy to use or manipulate

I think this matches the description. It is not supposed to be defined by the user though. Rather they should receive a finished object from one of the functions of `MarkovChain`. 

> Do CommunicationClass intended to be manipulated outside of communication_classes function?

They should be most widely used there but there is a public method to access them. It should be read-only for the user but its creation needs it to be mutable. The goal was to let the user do something like `X.communication_class()` and out pops something that looks like the solutions to [this pdf](https://drive.google.com/file/d/1mE9IgOdO-Nu9-xAaei74n28v5jVD7yvW/view?usp=sharing).

Should I do something like:
```python
class DiscreteMarkovChain(DiscreteTimeStochasticProcess, MarkovProcess):
    class CommunicationClass:
        ...
    ...
```

> like caching the CommunicationClass objects

Sadly, I have no clue on how to do that. I only know the `@cacheit` decorator for functions but that's about all.

Reply:
> If you partition the matrix into strongly connected components then I think that those are the communication classes:

Looks like it to me.

However, my function (I hope) also works with a wider range of matrices than just a numeric Matrix object. Also, it finds deeper properties such as whether those states are recurrent or transient and also their periodicity.

Reply:
The `strongly_connected_components` function can work with pretty much anything as the vertices of the graph. The only requirement is hashability.

I think that we should be trying to reuse code from other parts of sympy. If those functions aren't suitable for this case then maybe they can be improved.

What do recurrence, transience and periodicity mean in terms of the graph? I'd guess that a transient communication class is one that is not a leaf node of the DAG and a recurrent class is one that is. Those are straight-forward to find if you split into weakly and then strongly connected components.

Reply:
> I'd guess that a transient communication class is one that is not a leaf node of the DAG and a recurrent class is one that is.

I think that's correct.

> What does periodicity mean in terms of the graph?

I'm not sure how to express it in terms of graph theory (since I haven't studied it). But a state `i` is defined to be periodic with period `d` iff a return to that state is only ever possible in a number of steps that is a multiple of `d`. That is, `d=gcd{p_{ii}^{(n)} > 0}` where `p_{ii}^{(n)}` is the probably of returning to state `i` in n steps.

Reply:
I have a better draft about how to design the class as sympy objects, but I'd better accept any suggestions to make this to solve problems in more atomic fashion.
I don't know if `period` and `recurrent` are really everything about the communication classes, but if it turns out to be more problems to solve than this, I find that that it can be difficult to extend the old designs.

```diff
diff --git a/sympy/stats/stochastic_process_types.py b/sympy/stats/stochastic_process_types.py
index b4c54340ca..ddbe436fcf 100644
--- a/sympy/stats/stochastic_process_types.py
+++ b/sympy/stats/stochastic_process_types.py
@@ -583,7 +583,7 @@ def expectation(self, expr, condition=None, evaluate=True, **kwargs):


 class CommunicationClass(Basic):
-    def __new__(cls, states: list, recurrent: bool = None, period: int = None):
+    def __new__(cls, states, recurrent, period):
         """Represents a single communication class for a Markov Chain.

         Parameters
@@ -596,8 +596,28 @@ def __new__(cls, states: list, recurrent: bool = None, period: int = None):
         period: int, optional
             The period of the states. None if it is transient.
         """
+        states = _sympify(tuple(states))
+        recurrent = _sympify(recurrent)
+        period = _sympify(period)
+        return Basic.__new__(cls, states, recurrent, period)
+
+    @property
+    def states(self):
+        return self.args[0]
+
+    @property
+    def recurrent(self):
+        return self.args[1]
+
+    @property
+    def period(self):
+        return self.args[2]
+
+
+class _CommunicationClass:
+    def __new__(cls, states: list, recurrent: bool = None, period: int = None):
         # shamefully stolen from MutableDenseMatrix
-        self = Basic.__new__(cls)
+        self = object.__new__(cls)
         self.states = states
         self.recurrent = recurrent
         self.period = period
@@ -619,7 +639,7 @@ def __add__(self, other):
         else:
             period = None
         states.sort()
-        return CommunicationClass(states, recurrent, period)
+        return _CommunicationClass(states, recurrent, period)

     def __contains__(self, item):
         """Checks if a state is in the communication class"""
@@ -1075,7 +1095,7 @@ def communication_classes(self):
             P_prime[diag, diag] = 1

         Pn = P_prime ** n
-        classes = [CommunicationClass([i], True, periods[i]) for i in range(n)]
+        classes = [_CommunicationClass([i], True, periods[i]) for i in range(n)]
         for row in range(0, n):
             for col in range(row + 1, n):
                 # we use != instead of > in order to deal with symbols
@@ -1094,7 +1114,12 @@ def communication_classes(self):
             if i not in to_remove:
                 new_classes.append(classes[i])
         classes = new_classes
-        return classes
+
+        new_classes = []
+        for c in classes:
+            c = CommunicationClass(c.states, c.recurrent, c.period)
+            new_classes.append(c)
+        return new_classes

     def canonical_form(self):
         """
```

Reply:
Added. Thank you for the advice. There are very likely to be more interesting aspects to communication classes than just recurrence and periodicity.

I'll keep this thread unresolved in case anyone still has any comments.

Reply:
As you may notice, in sympy, classes are designed to have significant meanings and it is heavily going to be used inside another sympy objects.

The only point I’d not be happy about is introducing objects for return
It would be better just to unpack the classes and return everything in native python objects if they are not going to be used anywhere else

e.g. collection of tuples of (states, is_recurrent, period) or such

Reply:
```
diff --git a/sympy/stats/stochastic_process_types.py b/sympy/stats/stochastic_process_types.py
index b4c54340ca..da7b51b0ba 100644
--- a/sympy/stats/stochastic_process_types.py
+++ b/sympy/stats/stochastic_process_types.py
@@ -965,14 +965,14 @@ def limiting_dist(self, p0=None):
         comm_classes = self.communication_classes()

         # if irreducible and aperiodic
-        if (len(comm_classes) == 1) and (comm_classes[0].period == 1):
+        if (len(comm_classes) == 1) and (comm_classes[0][2] == 1):
             return self.stationary_distribution()

         # limits do not work well with matrices
         # I think this is still the stationary distribution
         # if all recurrent classes are aperiodic.
         # I have no sources to back this up though.
-        if all([c.period == 1 for c in comm_classes if c.recurrent]):
+        if all([c[2] == 1 for c in comm_classes if c[1] == True]):
             return self.stationary_distribution()

         # If the chain is reducible or periodic, the
@@ -1094,7 +1094,11 @@ def communication_classes(self):
             if i not in to_remove:
                 new_classes.append(classes[i])
         classes = new_classes
-        return classes
+
+        result = []
+        for c in classes:
+            result.append((c.states, c.recurrent, c.period))
+        return result

     def canonical_form(self):
         """
@@ -1166,10 +1170,10 @@ def canonical_form(self):
         r_states = []
         t_states = []
         for c in classes:
-            if c.recurrent:
-                r_states += c.states
+            if c[1]:
+                r_states += c[0]
             else:
-                t_states += c.states
+                t_states += c[0]

         states = r_states + t_states
         P_new = Matrix(n, n, lambda i, j: trans_probs[states[i], states[j]])
@@ -1259,10 +1263,10 @@ def decompose(self):
         r_states = []
         t_states = []
         for c in classes:
-            if c.recurrent:
-                r_states += c.states
+            if c[1]:
+                r_states += c[0]
             else:
-                t_states += c.states
+                t_states += c[0]

         states = r_states + t_states

diff --git a/sympy/stats/tests/test_stochastic_process.py b/sympy/stats/tests/test_stochastic_process.py
index 47956d8337..4e58956d80 100644
--- a/sympy/stats/tests/test_stochastic_process.py
+++ b/sympy/stats/tests/test_stochastic_process.py
@@ -107,9 +107,8 @@ def test_DiscreteMarkovChain():
     p = Y3.stationary_distribution()
     l = Y3.limiting_distribution
     assert p * TO3 == p == l
-    assert Y2.communication_classes() == [CommunicationClass([0], True, 1),
-                                          CommunicationClass([1, 2], False, 1)]
-    assert Y3.communication_classes() == [CommunicationClass([0, 1, 2], True, 1)]
+    assert Y2.communication_classes() == [([0], True, 1), ([1, 2], False, 1)]
+    assert Y3.communication_classes() == [([0, 1, 2], True, 1)]
     assert Y2.canonical_form() == ([0, 1, 2], TO2)
     assert Y3.canonical_form() == ([0, 1, 2], TO3)
     assert Y2.decompose() == ([0, 1, 2], TO2[0:1, 0:1], TO2[1:3, 0:1], TO2[1:3, 1:3])
(END)
```
My point can be this if we are not sure if the new classes with its own algebra, knowledge should be supplied for users.